[ExternalNode] Handle ExternalNode from Antrea agent side

mengdie-song opened the pull request:
1. Provide an example RBAC yaml file for Antrea agent running
on VM with definitions of ClusterRole, ServiceAccount and
ClusterRoleBinding.

2. Add ExternalNodeController to monitor ExternalNode CRUD,
invoke interfaces to operate OVS and update interface store
with ExternalEntityInterface.

3. Implement OVS interactions related to ExternalNode CRUD.

4. Add a channel for receiving ExternalEntity updates from
ExternalNodeController and notifying NetworkPolicyController
to reconcile rules related to the updated ExternalEntities.
This is to handle the case when NetworkPolicyController reconciles
rules before ExternalEntityInterface is realized in the
interface store.

5. Support configuring policy bypass rules to skip ANP check.

Signed-off-by: Mengdie Song <songm@vmware.com>
Co-authored-by: Wenying Dong <wenyingd@vmware.com>

mengdie-song commented:
/test-all

wenyingd commented on pkg/config/agent/config.go:
```diff
@@ -197,6 +197,10 @@ type AgentConfig struct {
+	// The namespace to create the ExternalNode for the VM/BM object.
+	// The default value is "default".
+	// It is used only when NodeType is externalNode.
+	Namespace string `yaml:"namespace,omitempty"`
```
Add the configuration item into path `build/yamls/externalnode/conf/antrea-agent.conf`

tnqn commented on pkg/agent/controller/networkpolicy/cache.go:
```diff
@@ -379,6 +388,31 @@ func (c *ruleCache) processPodUpdate(e interface{}) {
 }
 
+// processExternalNodeUpdate will be called when ExternalNodeController publishes an ExternalNode update event.
+// It finds out AppliedToGroups that contains this ExternalNode converted ExternalEntity and trigger reconciling
```
```suggestion
// It finds out AppliedToGroups that contains this ExternalNode converted ExternalEntity and triggers reconciling
```

tnqn commented on pkg/agent/controller/networkpolicy/cache.go:
```diff
@@ -379,6 +388,31 @@ func (c *ruleCache) processPodUpdate(e interface{}) {
+		return false
+	}
+	for group, memberSet := range c.appliedToSetByGroup {
+		for _, member := range memberSet.Items() {
```
can it use `memberSet.Has` like `processPodUpdate`? will members in `c.appliedToSetByGroup` have IPs?

mengdie-song replied:
Thanks for the suggestion, let me check this part further to see if memberSet.Has can work. Do you mean if interface contains IPs in this case? Yes, ExternalNodeController will notify NetworkPolicyController after the ExternalNode is created or updated and the IPs exist for this interface in the interfaceStore.

tnqn replied:
If the member in `c.appliedToSetByGroup` has IPs set, `memberSet.Has` cannot be used to query because when members are stored, their keys are populated based on name+namespace+ips, but when we query it, we don't have IPs, so keys will be different.

mengdie-song replied:
Thanks for the information! For ExternalEntity, controller will populate IPs for group member https://github.com/antrea-io/antrea/blob/main/pkg/controller/networkpolicy/networkpolicy_controller.go#L1197.
I think that name and namespace is enough for ExternalEntity case, so I revert the code to the first version.

mengdie-song replied:
Have updated by notifying ExternalEntityRef.

tnqn commented on pkg/agent/controller/networkpolicy/reconciler.go:
```diff
@@ -873,22 +873,22 @@ func (r *reconciler) GetRuleByFlowID(ruleFlowID uint32) (*types.PolicyRule, bool
 	for _, m := range members {
-		var entityName, ns string
 		var ifaces []*interfacestore.InterfaceConfig
+		var name, ns string
```
why changing this?  entityName represents this is the name of an entity, either a Pod or an ExternalEntity.

Anandkumar26 commented on pkg/agent/agent.go:
```diff
@@ -1153,6 +1169,15 @@ func (i *Initializer) initNodeLocalConfig() error {
+		}
+		return true, nil
+
+	}, i.stopCh)
```
Prefer to add a log message here as the process will be stuck in this state if no externalnode is created.

jianjuns commented on build/yamls/externalnode/conf/antrea-agent.conf:
```diff
@@ -32,6 +32,10 @@ featureGates:
 # Defaults to "k8sNode". Valid values include "k8sNode", and "externalNode".
 nodeType: externalNode
 
+# Namespace is the expected namespace to create the ExternalNode for the VM/BM object.
```
"the expected Namespace in which the ExternalNode should be created for a VM or baremetal server Node"

jianjuns commented on build/yamls/externalnode/conf/antrea-agent.conf:
```diff
@@ -32,6 +32,10 @@ featureGates:
 
+# Namespace is the expected namespace to create the ExternalNode for the VM/BM object.
+# Defaults to "default".
+#namespace: default
```
externalNodeNamespace?

jianjuns commented on build/yamls/externalnode/vm-agent-rbac.yml:
```diff
@@ -0,0 +1,101 @@
+kind: ServiceAccount
+metadata:
+  name: vm-agent
+  namespace: test-ns # Change the namespace to where vm-agent is expected to run.
```
If we will have a document to run VM agent with the sample yamls, I suggest to use a name that can be used in the installation guide without changing the yaml, e.g. "vm-ns".

jianjuns replied:
namespace -> Namespace

jianjuns commented on build/yamls/externalnode/vm-agent-rbac.yml:
```diff
@@ -0,0 +1,101 @@
+  - apiGroups:
+      - crd.antrea.io
+    resourceNames:
+      - VM1 # Change the ExternalNode name which vm-agent is expected to update.
```
Consider "vm-1".

jianjuns commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +241,172 @@ func DeleteOVSPort(brName, portName string) error {
+	pollErr := wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
+		renameErr = renameHostInterface(from, to)
+		if renameErr != nil {
+			klog.InfoS("Unable to rename host interface name with error, retrying", "oldName", from, "newName", to, "err", renameErr)
```
Maybe remove "with error".

mengdie-song replied:
OK, got it.

jianjuns commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +241,172 @@ func DeleteOVSPort(brName, portName string) error {
+	if err := ConfigureLinkAddresses(toIndex, ips); err != nil {
+		return err
+	}
+	// Copy the uplink interface's Route to the veth interface.
```
Route -> routes

jianjuns commented on build/charts/antrea/conf/antrea-agent.conf:
```diff
@@ -53,9 +53,6 @@ featureGates:
 # Enable certificated-based authentication for IPsec.
 {{- include "featureGate" (dict "featureGates" .Values.featureGates "name" "IPsecCertAuth" "default" false) }}
 
-# Enable running agent on an unmanaged VM/BM.
```
Ok. Please also update antrea-feature.go and features.md if needed.

jianjuns commented on pkg/ovs/ovsctl/appctl.go:
```diff
@@ -227,3 +228,45 @@ func (c *ovsCtlClient) GetDPFeatures() (map[DPFeature]bool, error) {
 	return features, nil
 }
+
+func (c *ovsCtlClient) DeleteDPInterface(name string) error {
```
Probably add a comment saying if the interface does not exists, the func just returns with no error.

mengdie-song commented on build/yamls/externalnode/vm-agent-rbac.yml:
```diff
@@ -0,0 +1,102 @@
+  - apiGroups:
+      - crd.antrea.io
+    resourceNames:
+      - vm-1 # Change the ExternalNode name which vm-agent is expected to update.
```
Hi @jianjuns @tnqn @wenyingd,

I am not sure if we prefer to specify resourceNames value here or not. 
If the rbac is used by one vm, then it is fine.
But in the cloud setup, the file will be shared for a VPC and the list will be huge.
Maybe we should remove this? I would like to know you idea, thanks!

wenyingd replied:
I agree to remove the `resourceNames` here, as it is actually inside ClusterRole definitation, and the ClusterRole actually shares among all VPCs. For per VPC, the ClusterRoleBinding is different as the ClusterRole is expected to bind the ServiceAccount in its namespace.

jianjuns replied:
Sounds reasonable to me.

jianjuns replied:
Or, we can add a comment saying resourceNames can be used to restrict access to only the current Node.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+		ovsExternalIDHostIFName:               port.ExternalIDs[ovsExternalIDHostIFName],
+		ovsExternalIDUplinkPort:               port.ExternalIDs[ovsExternalIDUplinkPort],
+		ovsExternalIDEntityName:               eeName,
+		ovsExternalIDEntityNamespace:          port.ExternalIDs[ovsExternalIDEntityName],
```
Agreed. Since namespace is tied to the antrea-agent.conf, it wont change during runtime. 
The reason I asked this, windows agent was relying on this function to update ovsExternalIDEntityName and ovsExternalIDEntityNamespace, that needs to be changed now.

Anandkumar26 replied:
We can update the namespace in the windows initializer. But not the ExternalEntity name
Do you mean, it should generate the name based of ExternalNode name and interface name?

wenyingd replied:
> We can update the namespace in the windows initializer. But not the ExternalEntity name Do you mean, it should generate the name based of ExternalNode name and interface name?

Yes, I think we have enough information to do it in initializer directly.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+		return nil
+	}
+	klog.InfoS("Deleting interface", "ifName", ifName)
+	if err := c.removeOVSPortsAndFlows(hostIface); err != nil {
```
As part of deleting externalnde, agent is restoring the interface configuration back to Host VM. 
Should we stop/restart antrea-agent for both Linux and Windows, to keep the end-end workflow same?

jianjuns replied:
But how you plan to restart agent? Should that be a manual step by users, or you mean the agent should exit itself? I feel the latter is not a good way.

wenyingd replied:
We provide a sample to register antrea-agent as a system service on both Linux and Windows. If agent process is quitted but the service itself is not stopped, the antrea-agent process woule be restarted automatically by the service.

But for the scenario that no system service is registered, user should restart the process by himself.

jianjuns replied:
I just feel it is strange the agent process exists at config change, and relying on another solution to restart it. Probably we should do that on Linux. If we need to do that on Windows, please add clear log message before exiting.

mengdie-song replied:
@tnqn @jianjuns I would like to confirm that if you think we should also stop agent for Linux to be consistent with Windows? For now, I do not let agent restart for Linux. I would change this part if you think the behavior should be consistent for these two operating systems.

mengdie-song replied:
Have confirmed with Jianjun and Quan and we prefer to keep existing behavior and do not let agent exit for Linux.

wenyingd commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	return nil
+}
+
+func (c *ExternalNodeController) reconcileExternalEntityInterfaces() error {
```
Rename as "reconcileExternalNodeInterfaces"?

wenyingd commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +241,172 @@ func DeleteOVSPort(brName, portName string) error {
+
+func RenameInterface(from, to string) error {
+	var renameErr error
+	pollErr := wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
```
Add comment to explain that adding retry here because it might return `device is busy` error for one time configuration.

wenyingd commented on pkg/ovs/ovsctl/appctl.go:
```diff
@@ -227,3 +228,46 @@ func (c *ovsCtlClient) GetDPFeatures() (map[DPFeature]bool, error) {
 	return features, nil
 }
+
+// DeleteDPInterface deletes DP interface, and it returns with no error if the interface does not exist.
```
```suggestion
// DeleteDPInterface deletes OVS datapath interface, and it returns with no error if the interface does not exist.
```

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	if err != nil {
+		return nil, err
+	}
+	attachInfo := getOVSAttachInfo(uplinkName, uplinkUUID, fmt.Sprintf("%d", hostIF.Index), hostIF.Name, eeName, eeNamespace, ips)
```
It's not a must for windows. I have followed the same steps as linux to get windows agent working. 
If index is not needed anywhere, then it is not required to index field.

mengdie-song replied:
I removed the index in the latest change.  Also, @Anandkumar26  Could you help confirm if ovsExternalIDHostIFName in the ExternalIDs is needed for windows since it is same as InterfaceName on Linux. If windows doesn't need it, I would like to remove this one.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+		return nil
+	}
+	klog.InfoS("Updating OVS port data", "ifName", ifName, "eeName", eeName, "ip", ips)
+	iface, err := c.updateOVSPortsData(hostIface, eeName, ips)
```
We don't have the externalEntity name available in the initializer. So windows can only set the namespace field.
Windows rely on `updateOVSPortsData` to update the eeName in the ovs port.

mengdie-song replied:
updateOVSPortsData will be called when either ExternalEntity name or interface ips are changed. If neither of them is changed, the logics should not go into this part.

mengdie-song replied:
Although we will only step into updateOVSPortsData when either ExternalEntity name or interface ips are changed. But if windows choose to create the port with all the needed fields including ExternalEntity name, we will have need this check for unnecessary update. So I update this part, please help review the latest version.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalNodeListerSynced) {
+		klog.Error("Failed to wait for syncing cache for ExternalNodes")
```
Failed to wait for syncing ExternalNodes cache?

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
```
nit: Expected string type

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncExternalNode(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
```
If no error occurs, then forget this item so it does not get queued again until

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
```
Pls keep the name consistent ("workqueue" or "work queue")

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+		return nil, err
+	}
+	if err := util.RenameInterface(hostIfName, uplinkName); err != nil {
+		klog.ErrorS(err, "Failed to rename host interface name as the uplink name", "hostInterface", hostIfName, "uplink", uplinkName)
```
Failed to rename host interface name to uplink name

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	}()
+	hostOFPort, err := c.ovsBridgeClient.GetOFPort(hostIfName, false)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get host interface ofport", "hostInterface", uplinkName)
```
klog.ErrorS(err, "Failed to get host interface ofport", "hostInterface", hostIfName)

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	if err := c.ofClient.InstallVMUplinkFlows(hostIfName, hostOFPort, uplinkOFPort); err != nil {
+		return nil, err
+	}
+	klog.InfoS("Added uplink and host port on OVS and installed openflow entries", "uplink", uplinkName, "hostInterface", hostIfName)
```
nit: Added uplink and host port in OVS

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+		return err
+	}
+	if err := c.ovsBridgeClient.DeletePort(uplinkPortID); err != nil {
+		return err
```
Can we add log message here similar to L655

mengdie-song replied:
Sure, I have updated this part.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+}
+
+// Run will create defaultWorkers workers (goroutines) which will process the ExternalNode events from the work queue.
+func (c *ExternalNodeController) Run(stopCh <-chan struct{}) {
```
Can we organize functions such that all the public functions for a given struct are at the top, while the private functions are after the public functions?

Public Functions
- NewExternalNodeController
- Run
Private Functions 
- enqueueExternalNodeAdd
- enqueueExternalNodeUpdate

Other functions which are not part of the struct, can be placed at the end of file.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+
+func getHostInterfaceName(iface v1alpha1.NetworkInterface) (string, error) {
+	var ipFilter *ip.DualStackIPs
+	epIP := net.ParseIP(iface.IPs[0])
```
Is it necessary to have this function in externalnode package? Seems util package is the appropriate place

mengdie-song replied:
Thanks for pointing out. I have changed to process all the IPs. I did not move this to util because it is not used by other packages, but only this file.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	}
+	_, _, link, err := util.GetIPNetDeviceFromIP(ipFilter, sets.NewString())
+	if err == nil {
+		klog.InfoS("Using the interface", "linkName", link.Name)
```
This is generating multiple log messages. (once for ADD operation and twice for UPDATE event).

I0624 05:58:56.480822    4680 external_node_controller.go:718] "Using the interface" linkName="Ethernet"
I0624 05:58:56.480822    4680 external_node_controller.go:382] "Updating OVS port data" ifName="Ethernet" eeName="win2019-23946" ip=[10.110.0.5]
I0624 05:58:56.485247    4680 external_node_controller.go:248] "Installed reserved flows"
I0624 05:58:56.485247    4680 external_node_controller.go:200] ExternalNodeController finishes reconciliation
I0624 05:58:56.485247    4680 external_node_controller.go:392] "Updating ExternalNode" ExternalNode="vm-ns/win2019"
I0624 05:58:56.499223    4680 external_node_controller.go:718] "Using the interface" linkName="Ethernet"
I0624 05:58:56.514986    4680 external_node_controller.go:718] "Using the interface" linkName="Ethernet"

mengdie-song commented on pkg/agent/controller/networkpolicy/cache.go:
```diff
@@ -379,6 +388,28 @@ func (c *ruleCache) processPodUpdate(e interface{}) {
+// of related rules.
+// It can enforce NetworkPolicies to ExternalEntities after ExternalEntityInterface is realised in the interface store.
+func (c *ruleCache) processExternalNodeUpdate(e interface{}) {
+	externalNode := e.(*crdv1alpha1.ExternalNode)
```
Sure, will update.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	nodeName           string
+	namespace          string
+	reservedHostPorts  []reserveHostPort
+	reservedRules      []string
```
Got it, will update.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+}
+
+func (c *ExternalNodeController) getReservedHostPorts() error {
+	for _, s := range c.reservedRules {
```
OK. We don't need to change reservedHostPorts, right?

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	}
+	c.syncedExternalNodes.Add(en)
+	// Notify the ExternalNode create event to NetworkPolicyController.
+	c.nodeUpdateNotifier.Notify(en)
```
Will update to notify ExternalEntityReference

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	hostIface, portExists := c.ifaceStore.GetInterfaceByName(ifName)
+	if !portExists {
+		klog.InfoS("Creating OVS ports and flows for ExternalEntityInterface", "ifName", ifName, "eeName", eeName, "ip", ips)
+		uplinkName := genUplinkInterfaceName(ifName)
```
Sure.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	return interfaceConfig, nil
+}
+
+func genUplinkInterfaceName(hostIfName string) string {
```
Sure

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+}
+
+func (c *ExternalNodeController) reconcile() error {
+	klog.Info("ExternalNodeController starts reconciliation")
```
Changed the log here, but I use Klog.InfoS instead.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+	if err := c.reconcilePolicyBypassFlows(); err != nil {
+		return fmt.Errorf("failed to reconcile reserved flows %v", err)
+	}
+	klog.Info("ExternalNodeController finishes reconciliation")
```
I prefer to keep it here for easy debugging. It is used to distinguish reconcile process between normal events.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+		return err
+	}
+	if err := c.ovsBridgeClient.DeletePort(port.UUID); err != nil {
+		klog.ErrorS(err, "Failed to port in OVS", "hostInterface", hostIfName)
```
nit: Failed to delete

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,765 @@
+	klog.InfoS("Updating ExternalNode", "ExternalNode", klog.KObj(curEn))
+	preEn := obj.(*v1alpha1.ExternalNode)
+	if reflect.DeepEqual(preEn.Spec.Interfaces[0], curEn.Spec.Interfaces[0]) {
+		klog.InfoS("Skipping processing ExternalNode update as no changes for Interface[0]", "ExternalNode", klog.KObj(curEn))
```
nit: Skip processing

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,765 @@
+		klog.InfoS("Skipping processing ExternalNode update as no changes for Interface[0]", "ExternalNode", klog.KObj(curEn))
+		return nil
+	}
+	preEEName := externalnode.GenExternalEntityName(preEn)
```
Pls rebase.  This patch breaks compilation on feature/externalnode.

jianjuns commented on pkg/config/agent/config.go:
```diff
@@ -204,6 +204,16 @@ type AgentConfig struct {
+	// The default value is "default".
+	// It is used only when NodeType is externalNode.
+	ExternalNodeNamespace string `yaml:"externalNodeNamespace,omitempty"`
+	// The bypass policy rules are used to skip the check from Antrea NetworkPolicy rules.
```
"rules define traffic that should bypass NetworkPolicy rules"?

jianjuns commented on pkg/config/agent/config.go:
```diff
@@ -204,6 +204,16 @@ type AgentConfig struct {
+	// direction (in|out), protocol(tcp/udp/icmp), remote IP or CIDR, dst port (ICMP doesn't require), and use ":" to
+	// split the items, e.g., "in:tcp:1.1.1.1:22"
+	// It is used only NodeType is externalNode.
+	BypassPolicyRules []string `yaml:"bypassPolicyRules,omitempty"`
```
Should we name it "PolicyBypassRules" or "PolicyBypassingRules"

mengdie-song replied:
Thanks, I have changed the field to "PolicyBypassRules".

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+
+func (c *ExternalNodeController) getReservedHostPorts() error {
+	for _, s := range c.bypassPolicyRules {
+		// Each rule contains the following four items.
```
The same comment.

jianjuns commented on pkg/agent/controller/networkpolicy/cache.go:
```diff
@@ -443,6 +451,33 @@ func (c *ruleCache) processPodUpdate(e interface{}) {
 }
 
+// processExternalNodeUpdate will be called when ExternalNodeController publishes an ExternalNode update event.
+// It finds out AppliedToGroups that contains this ExternalNode converted ExternalEntity and triggers reconciling
```
Sorry, here I actually meant changing "contains" to "contain", not changing "AppliedToGroups".

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+	hostIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.ExternalEntityInterface)
+	for _, hostIface := range hostIfaces {
+		if err := c.ofClient.InstallVMUplinkFlows(hostIface.InterfaceName, hostIface.OVSPortConfig.OFPort, hostIface.UplinkPort.OFPort); err != nil {
+			klog.ErrorS(err, "Failed to install openflow entries to forward packet between uplink and host interface", "hostInterface", hostIface.InterfaceName)
```
packet -> packets

But I feel you can simply say "Failed to install host uplink flows".

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+		return err
+	}
+	for _, rhp := range c.reservedHostPorts {
+		klog.V(2).InfoS("Installing reserved flows", "protocol", rhp.protocol, "IP", rhp.ip, "IPNet", rhp.ipnet, "port", rhp.port, "ingress", rhp.ingress)
```
reserved flows -> policy bypassing flows

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+	return c, nil
+}
+
+// Run will create defaultWorkers workers (goroutines) which will process the ExternalNode events from the work queue.
```
It runs only a single worker?

mengdie-song replied:
I will update the comment here. Since agent will only process its ExternalNode, we think one worker is enough here.

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+	preEEName := portData.ExternalIDs[ovsExternalIDEntityName]
+	preIPs := sets.NewString(strings.Split(portData.ExternalIDs[ovsExternalIDIP], eeSplitter)...)
+	if (preEEName == eeName) && sets.NewString(ips...).Equal(preIPs) {
+		klog.InfoS("Skipping updating OVS port data as both entity-namespace and ip are not changed", "ifName", ifName)
```
I did not see you check entity-namespace. Do you mean name?

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+	return nil
+}
+
+func (c *ExternalNodeController) createOVSPortsAndFlows(uplinkName, hostIfName, eeNamespace string, eeName string, ips []string) (*interfacestore.InterfaceConfig, error) {
```
hostIfName -> hostIFName

I think we use IF in most code.

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+	}
+	defer func() {
+		if !success {
+			c.ovsBridgeClient.DeletePort(uplinkUUID)
```
Log the error if any of rollback funcs fails.

And what will happen if rollback fails? Will the retry or deletion be able to resume the creation or clean up the state?

mengdie-song replied:
I have added the logs here as suggested. Since it will return error to its caller which finally is function syncExternalNode, I think it should be put back to the work queue and retry.

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+func (c *ExternalNodeController) createOVSPortsAndFlows(uplinkName, hostIfName, eeNamespace string, eeName string, ips []string) (*interfacestore.InterfaceConfig, error) {
+	adapterConfig, err := util.GetUplinkConfig(hostIfName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get the configuration on the host interface", "hostInterface", hostIfName)
```
Paste my previous comment to another PR:

> For all the error logs, could you check we do need to log them, or we can append the error message to the returned error, and log in the caller? It is ok in some cases you like to log twice with more information where the error first happens, but we need to avoid many duplicated error logs, and esp. if you just return the error without prepending extra context from the callee, it may generate confusing log in the caller (as the caller usually logs the returned error).
> 
> For example, assume func A calls B calls C, and C gets a "not-found" error. If all A, B, C log the error and directly return the error to the caller, we will get: "A: not found", "B: not found", "C: sth is not found". First, it is unnecessary to have the same error logged three times; second, just "not-found" without further information could be confusing in the logs in A and B. So, we usually prepend the error context to the returned error and log in the caller, e.g. in C - return fmt.Errorf("failed to do sth: %v", err)

wenyingd commented on pkg/agent/agent.go:
```diff
@@ -280,9 +288,17 @@ func (i *Initializer) initInterfaceStore() error {
-				// specific interface in the interfaceStore.
-				intf = nil
+				if port.Name == i.ovsBridge {
+					// Not load the host interface, because it is configured on the OVS bridge port, and we don't need a
```
By now, OVS bridge port is mainly used in container case ( for Windows or Antrea IPAM), and no consumer on the OVS bridge interfaceConfig exists, instead, we always use `AdapterNetConfig` to save the uplink configuration in `NodeConfig`. So we didn't load it to any InterfaceConfig. 

For VM case, OVS bridge doesn't  have an OVS port or interface, even mapping uplink. 

Since we want to unify the implementation on AntreaIPAM/Windows/VM Support, I plan to remove the OVS bridge port in a future patch, then this logic should be removed.

jianjuns replied:
I feel even today, we need not to set `AntreaHost` to the bridge port?

wenyingd replied:
No. If we don't set any interfaceType on the OVS bridge port, the logic would think go into the original logic ( not setting OVS port type), and in this case we can not differentiate the container port or OVS bridge port because there is no identifiers. Then the OVS bridge port is thought as container InterfaceConfig added into the interface store by this line:
```
			switch {
			...
			default:
				// The port should be for a container interface.
				intf = cniserver.ParseOVSPortInterfaceConfig(port, ovsPort, true)
				antreaIFType = interfacestore.AntreaContainer
			}
```

jianjuns replied:
I thought we can just use bridge name to recognize the bridge port. At least it should not be set with AntreaHost, esp. after we change host interface to a separate internal port.

wenyingd replied:
Do you mean change the logic to identify Bridge port via name in this change? After we change host interface to a separate internal port, we do not need Bridge port any more, so it should not exist.

wenyingd replied:
> We can change later. To support upgrade, we may still need to identify and ignore bridge port?

Yes, you are correct

wenyingd commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +241,173 @@ func DeleteOVSPort(brName, portName string) error {
 	return cmd.Run()
 }
+
+func GetNetRoutesOnAdapter(linkIndex int) (string, []interface{}, error) {
```
Rename as "GetRoutesAndDefaultGwOnInterface"?

wenyingd commented on pkg/agent/util/net.go:
```diff
@@ -389,3 +389,23 @@ func GenerateRandomMAC() net.HardwareAddr {
+	var addrs []*net.IPNet
+	for _, a := range addrList {
+		if ipNet, ok := a.(*net.IPNet); ok {
+			if ipNet.IP.IsGlobalUnicast() {
```
Maybe we should not filter only global IP?

mengdie-song replied:
I updated this part by removing this check. Instead, I modify ConfigureLinkAddresses in net_linux.go to make sure it will not return error if the address has already existed. Please help review the latest the version.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+			},
+		}
+	}
+	interfaceConfig.EntityInterfaceConfig = hostUplinkConfig
```
This will be set to an empty value when there is no ExternalIDs set for the uplink port, with no error returned to the called.

Anandkumar26 commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+			return err
+		}
+	}
+	klog.InfoS("Installed reserved flows")
```
klog.Info?

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,753 @@
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeAdd(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
```
Thanks for the comments. I tested that NewFilteredExternalNodeInformer can achieve this. Will update this part.

mengdie-song commented on build/yamls/externalnode/vm-agent-rbac.yml:
```diff
@@ -0,0 +1,100 @@
+    verbs:
+      - get
+      - watch
+      - list
```
I think so. Although I have already changed the code to use filtered informer and let the agent to watch its own ExternalNode, we still use the lister and we will get errors if I removed this privilege

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+	}
+	hostIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.ExternalEntityInterface)
+	for _, hostIface := range hostIfaces {
+		if hostIface.EntityName != eeName || hostIface.EntityNamespace != en.Namespace {
```
I am thinking that agent may change its configured externalNodeNamespace and we will need to clean up the stale one during reconciliation. But it is true that the case is really rare.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,777 @@
+		// Delete host interface from OVS datapath if it is existing.
+		// This is to resolve an issue that OVS fails to remove the interface from datapath. It might happen because the interface
+		// is busy when OVS tries to remove it with the OVSDB interface deletion event.
+		if err := c.ovsctlClient.DeleteDPInterface(hostIfName); err != nil {
```
I have synced with Anand and he has confirmed that hostIfName is also equal to the OVS internal port name for windows. As a result, I did not change this part. Meanwhile, I removed ovsExternalIDHostIFName from externalIDs since it is the same as InterfaceName in InterfaceConfig

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,756 @@
+
+// policyBypassConfig is generated by policyBypassRule which contains
+// direction (in|out), protocol(tcp/udp/icmp), remote IP or CIDR, dst port.
+type policyBypassConfig struct {
```
Should we just name it policyBypassRule?

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,756 @@
+}
+
+// policyBypassConfig is generated by policyBypassRule which contains
+// direction (in|out), protocol(tcp/udp/icmp), remote IP or CIDR, dst port.
```
Maybe add "Traffic matching the rule should bypass NetworkPolicy enforcement".

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,756 @@
+	defer func() {
+		if !success {
+			if ovsErr = c.ovsBridgeClient.DeletePort(hostIfUUID); ovsErr != nil {
+				klog.ErrorS(err, "failed to host interface port", "portUUID", hostIfUUID, "hostIFName", hostIFName)
```
failed to delete

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,756 @@
+	klog.InfoS("Adding ExternalNode", "ExternalNode", klog.KObj(en))
+	eeName, err := externalnode.GenExternalEntityName(en)
+	if err != nil {
+		return err
```
Thanks for the information!

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,756 @@
+	success := false
+	defer func() {
+		if !success {
+			if err = util.RenameInterface(uplinkName, hostIFName); err != nil {
```
I understand your point and Wenying and I have had a discussion on it. However, we did not find a way to avoid this loop of failure if we did get error in the defer clause. Retry will always start from the beginning rather than continue from where it failed last time. So you are right that we can get problem when we retry RenameInterface from hostIFName to uplinkName if the rollback fails. In my point, we may need to manually recover the setup if it did occur.

jianjuns replied:
At least we should add a comment about this, and even better to include it in the log message, e.g. sth. like "Manual cleanup is required".

jianjuns replied:
BTW, I remember NodeRouteController save the processing state to a cache, so retry knows where to start. But I know it can be complex and am not saying you must do that for ExternalNode.

mengdie-song replied:
Thanks for the comment. I have added this to log message as suggested.

jianjuns commented on build/yamls/externalnode/conf/antrea-agent.conf:
```diff
@@ -32,6 +32,17 @@ featureGates:
+# Defaults to "default".
+#externalNodeNamespace: default
+
+# The policyBypassRules describes the traffic that expects to bypass NetworkPolicy rules.
```
is expected

jianjuns commented on pkg/agent/controller/networkpolicy/cache.go:
```diff
@@ -443,6 +451,33 @@ func (c *ruleCache) processPodUpdate(e interface{}) {
+	}
+	for group, memberSet := range c.appliedToSetByGroup {
+		for _, m := range memberSet.Items() {
+			if externalEntityEquals(m.ExternalEntity, externalEntityRef) {
```
I feel not necessary to define a func for this, but up to you.

mengdie-song replied:
Thanks for pointing out and I agree that this is not necessary. I have changed to use reflect.DeepEqual(m.ExternalEntity, externalEntityRef) to check this part.

jianjuns commented on pkg/agent/controller/networkpolicy/cache.go:
```diff
@@ -443,6 +451,33 @@ func (c *ruleCache) processPodUpdate(e interface{}) {
+	externalEntityRef := e.(v1beta.ExternalEntityReference)
+	c.appliedToSetLock.RLock()
+	defer c.appliedToSetLock.RUnlock()
+	externalEntityEquals := func(expEntity *v1beta.ExternalEntityReference, actEntity v1beta.ExternalEntityReference) bool {
```
Equals -> Matches?

jianjuns commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
+
+func RenameInterface(from, to string) error {
+	var renameErr error
+	// Add retry here because it might return device is busy error for one time configuration.
```
"the error of device is busy"?

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,757 @@
+		case "icmp":
+			proto = binding.ProtocolICMP
+		default:
+			proto = binding.ProtocolIP
```
I feel we should either define a name for IP or default to IP if no protocol is specified, but should return an error when validating the rules if not supported protocol is added.

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,757 @@
+		// parse port number.
+		var port int
+		if rule[3] != "" {
+			port, _ = strconv.Atoi(rule[3])
```
It can be invalid port number if we do not validate earlier

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,757 @@
+			if strings.Contains(rule[2], "/") {
+				_, remoteCIDR, err = net.ParseCIDR(rule[2])
+				if err != nil {
+					return rules, err
```
In general, I feel we should validate the rules in Options.validate() and even better convert them to go types there.

mengdie-song replied:
I have moved all the checks to validatePolicyBypassRulesConfig() in Options.validate(). It is to make sure that ExternalNodeController will not get invalid input. But I did not find a good way to do conversion at the same time under validation which can be used by ExternalNodeController, because validation part will only return an error. So I keep the conversion part here.

tnqn commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
+	iface, err := net.InterfaceByName(uplinkIfName)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get interface", "uplinkIfName", uplinkIfName)
+		return nil, err
```
Generally we should either return the error (with more context prepended) or log it but not do both of them, otherwise eventually it would cause duplicate errors, which may be confusing. This patch has a few such duplicate error handling.

tnqn commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
+	return nil
+}
+
+func MoveIFConfigurations(ips []*net.IPNet, routes []interface{}, mtu int, from string, to string) error {
```
As an util function, it's hard to understand what the function does from the arguments.
If it's a function moving ip configurations from one interface to another, there should be just two arguments represeting the source and destination.
If it's a function setting the provided ip address, routes, mtu to an interface, it doesn't make sense to pass `from` and have a logic to check the existence of `from`.
This business specific steps make the function cannot really be reused by other consumers and the function name may be confusing. I would suggest move some logic to caller side, like separating cleanup of `from` and configuration of `to`.

mengdie-song replied:
We put it here because windows will also need to invoke this function. But this function should do nothing for windows case. But it is true that the function is hard to reuse. So I move this function to external_node_controller_linux.go and external_node_controller_windows.go in the latest update. Please help review this part, thanks.

tnqn commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
+}
+
+// removeUplinkIPRoutes removes the IP and routes from the uplink interface.
+func removeUplinkIPRoutes(uplinkIF netlink.Link) error {
```
ditto, this is in util package, it shouldn't even care whether this is an unlink, and the implementation has nothing to do with uplink.
And is it really worth to have `removeUplinkIPRoutes` call `removeRoutesOnLink`? I feel just naming this function as `removeAddrsOnLink` and calling `removeAddrsOnLink` and `removeRoutesOnLink` is more clear.

tnqn commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
+	return nil
+}
+
+func copyRoutes(toIF netlink.Link, routes []interface{}) error {
```
ditto, this has nothing with "copy" and is just configuring routes on a link.

jianjuns commented on build/yamls/externalnode/conf/antrea-agent.conf:
```diff
@@ -32,6 +32,17 @@ featureGates:
+
+# The policyBypassRules describes the traffic that is expected to bypass NetworkPolicy rules.
+# Each rule contains the following four attributes:
+# direction (in|out), protocol(tcp/udp/icmp/ip), remote IP or CIDR, dst port (ICMP doesn't require),
```
Humm.. so you would use "ingress" / "egress"? I agree that is consistent with the NetworkPolicy definition.

mengdie-song replied:
Have changed to ingress/egress as suggested

jianjuns commented on cmd/antrea-agent/options.go:
```diff
@@ -496,9 +497,54 @@ func (o *Options) validateExternalNodeOptions() error {
+			return fmt.Errorf("missing port for policyBypassRule %s", rule)
+		}
+		if len(attributes) == 4 {
+			_, err := strconv.Atoi(attributes[3])
```
We should check the port number value falls into the valid range too.

mengdie-song replied:
Sure, I have added the check to make sure it ranges from 0 to 65535.

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,756 @@
+	ovsExternalIDUplinkPort      = "uplink-port"
+	ovsExternalIDEntityName      = "entity-name"
+	ovsExternalIDEntityNamespace = "entity-namespace"
+	ovsExternalIDIPs             = "ips"
```
A reminder for this one.

wenyingd replied:
The IPs are also used in `restart` cases. If we do not cache the IPs, it may cause the `InterfaceConfig` for the ExternalEntity which is loaded from OVSDB has no IPs, and the `reconcile` logic in network_policy_controller may failed to replay the OpenFlow entries on the egress rules because the IPs are not provided. This is because network_policy_controller and external_node_controller are different goroutines.

jianjuns replied:
Ok. I got confused why we specify the IPs in ExternalNode, but not discover them from interfaces? In the case of DHCP, users may not IPs when creating the ExternalNode.

wenyingd replied:
By now, we have validation on the existence of IP in Interfaces configuration when creating ExternalNode, although we support using both name and IPs in it. 

This additional check is from cloud VM, that the interface name is not known from the cloud API, but IP is always known. 

As for the case of DHCP, I didn't get your point why the user knows the interface name but does not know the IP.

jianjuns replied:
In the DHCP (and other auto-IPAM) case, users might not know the IP in advance, and the IP can change too (e.g. lease renew failed); interface name is static. I am not talking about public cloud VMs.

But I am fine to support only user specified IPs for now.

jianjuns commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+	defer func() {
+		if !success {
+			if err = util.RenameInterface(uplinkName, hostIFName); err != nil {
+				klog.ErrorS(err, "failed to restore uplink name back to host interface name, manual cleanup is required", "uplinkName", uplinkName, "hostIFName", hostIFName)
```
". Manual cleanup is required"

wenyingd commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+
+func (c *ExternalNodeController) reconcile() error {
+	klog.InfoS("Reconciliation for controller", "name", controllerName)
+	if err := c.reconcileHostUplinkFlows(); err != nil {
+		return fmt.Errorf("failed to reconcile host uplink flows %v", err)
+	}
+	if err := c.reconcileExternalNodeInterfaces(); err != nil {
```
`reconcileHostUplinkFlows` should be the first to ensure the datapath connectivity is  recovered. Then antrea-agent can get ExternalNode from APIServer.

tnqn replied:
If the datapath connectivity is not recovered, how antrea-agent gets the externalNode in `initVMLocalConfig` and gets here.

wenyingd replied:
`initVMLocalConfig` is called before `initFlows`, so OpenFlow entries are not flushed in the new round. The existing flows configured in last round supports the connectivity.

wenyingd commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+
+func (c *ExternalNodeController) reconcileExternalNodeInterfaces() error {
+	en, err := c.externalNodeLister.ExternalNodes(c.namespace).Get(c.nodeName)
+	if err != nil {
```
No, when logic comes into `ExternalNodeController`, it means the ExternalNode is already existing. The logic to wait for ExternalNode creation is in agent Initializer, which is before here.

tnqn replied:
I think it could happen an ExternalNode is removed after `initVMLocalConfig`, given the controller handles add/delete events. The controller could work if the ExternalNode is deleted after reconcilation but could not work if it is deleted in-between `initVMLocalConfig` and `reconcile`.

mengdie-song replied:
I have updated this part by checking the error type and now it will not return error if errors.IsNotFound returns true.

wenyingd commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+		// Delete host interface from OVS datapath if it exists.
+		// This is to resolve an issue that OVS fails to remove the interface from datapath. It might happen because the interface
+		// is busy when OVS tries to remove it with the OVSDB interface deletion event.
+		if err := c.ovsctlClient.DeleteDPInterface(hostIFName); err != nil {
```
Yes, it is hit in practice. The scene that we hit the error include: 1) the configurations for both uplink and internal port are removed in OVSDB; 2) the interface for the OVS internal port is removed on the host, 3) the uplink interface in OVS datapath is still attached. Since the ports in OVSDB is successfully removed, we hope that the interface in OVS datapath could be removed no matter the later steps are successful or not. So a `defer` is used here.

wenyingd commented on pkg/ovs/ovsctl/appctl.go:
```diff
@@ -227,3 +228,46 @@ func (c *ovsCtlClient) GetDPFeatures() (map[DPFeature]bool, error) {
+	}
+	scanner := bufio.NewScanner(strings.NewReader(string(out)))
+	scanner.Split(bufio.ScanLines)
+	ports := make([]int, 0)
```
No, there should be at most one port using the name.

But you do remind me, the existing logic uses `strings.Contains`, which might have issues if we have a second port name which is fully containing the name, e.g, on Windows we may have adapters like `Ethernet0` and `Ethernet0 2`. So we should modify the logic to find target port.

tnqn replied:
It seems the code still assume there could be more than one port?

mengdie-song replied:
I have changed this part, please help check the latest version. Thanks!

mengdie-song commented on build/yamls/externalnode/conf/antrea-agent.conf:
```diff
@@ -32,6 +32,17 @@ featureGates:
+# The policyBypassRules describes the traffic that is expected to bypass NetworkPolicy rules.
+# Each rule contains the following four attributes:
+# direction (in|out), protocol(tcp/udp/icmp/ip), remote IP or CIDR, dst port (ICMP doesn't require),
+# and these attributes are separated by ":", e.g. "in:tcp:1.1.1.1:22".
```
done

mengdie-song commented on build/yamls/externalnode/conf/antrea-agent.conf:
```diff
@@ -32,6 +32,17 @@ featureGates:
+# direction (in|out), protocol(tcp/udp/icmp/ip), remote IP or CIDR, dst port (ICMP doesn't require),
+# and these attributes are separated by ":", e.g. "in:tcp:1.1.1.1:22".
+# It is used only when NodeType is externalNode.
+#policyBypassRules:
```
done

mengdie-song commented on cmd/antrea-agent/agent.go:
```diff
@@ -60,6 +61,7 @@ import (
 	"antrea.io/antrea/pkg/agent/stats"
 	agenttypes "antrea.io/antrea/pkg/agent/types"
 	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions"
+	eninformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha1"
```
done

mengdie-song commented on cmd/antrea-agent/agent.go:
```diff
@@ -341,12 +354,21 @@ func run(o *Options) error {
+	} else {
+		gwPort = 0
+		tunPort = 0
+	}
```
done

mengdie-song commented on pkg/agent/agent.go:
```diff
@@ -1191,11 +1206,21 @@ func (i *Initializer) initNodeLocalConfig() error {
 
 func (i *Initializer) initVMLocalConfig(nodeName string) error {
+	klog.InfoS("Initializing VM config", "ExternalNode", nodeName)
+	wait.PollImmediateUntil(10*time.Second, func() (done bool, err error) {
```
Got it, done

mengdie-song commented on pkg/agent/config/node_config.go:
```diff
@@ -104,6 +104,8 @@ type AdapterNetConfig struct {
 	Index      int
 	MAC        net.HardwareAddr
 	IP         *net.IPNet
+	IPs        []*net.IPNet
```
I have removed the IP from AdapterNetConfig.

mengdie-song commented on pkg/agent/controller/networkpolicy/reject.go:
```diff
@@ -256,12 +260,19 @@ func getRejectOFPorts(rejectType RejectType, sIface, dIface *interfacestore.Inte
+		if dIface.Type == interfacestore.ExternalEntityInterface {
+			inPort = uint32(dIface.EntityInterfaceConfig.UplinkPort.OFPort)
+		} else {
+			inPort = config.HostGatewayOFPort
```
Right, it should be gwOFPort here. Have updated.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+	// The dst port of the given protocol.
+	port uint16
+	// IP is destination IP for Egress and source IP for Ingress.
+	ip net.IP
```
Thanks for the information. I have updated here, please review the latest version.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+// protocol(tcp/udp/icmp), remote IP or CIDR and dst port.
+// Traffic matching the rule should bypass NetworkPolicy enforcement
+type policyBypassRule struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match.
```
Thanks, done

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+	externalNodeListerSynced cache.InformerSynced
+	queue                    workqueue.RateLimitingInterface
+	ifaceStore               interfacestore.InterfaceStore
+	syncedExternalNodes      cache.Store
```
Thanks for pointing out, updated accordingly.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+}
+
+func (c *ExternalNodeController) reconcilePolicyBypassFlows() error {
+	rules, err := c.getPolicyBypassRules()
```
Since we have added validation,  it should be changed. Thanks!

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+		klog.V(2).InfoS("Installing policy bypass flows", "protocol", rhp.protocol, "IP", rhp.ip, "IPNet", rhp.ipnet, "port", rhp.port, "ingress", rhp.ingress)
+		if err = c.ofClient.InstallPolicyBypassFlows(rhp.protocol, rhp.ipnet, rhp.ip, rhp.port, rhp.ingress); err != nil {
+			klog.ErrorS(err, "Failed to install policy bypass flows", "protocol", rhp.protocol, "IP", rhp.ip, "IPNet", rhp.ipnet, "port", rhp.port, "direction", rhp.ingress)
+			return err
```
Thanks for pointing out. Do we need to check this error for InstallPolicyBypassFlows only? Or it should be checked for reconcileHostUplinkFlows() either? Also, does this error message have any keyword, like "OVS not ready"?

mengdie-song replied:
Got it, I have updated this part with `PollImmediate`. Do you think retry 10 times is enough? Please help check the latest verison.

mengdie-song replied:
Thanks a lot, Quan. I have updated this part according to our discussion, please help take a look at the latest version.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+	klog.InfoS("Added uplink port in OVS", "port", uplinkOFPort, "uplinkName", uplinkName)
+
+	// Create host port in OVS.
+	attachInfo := GetOVSAttachInfo(uplinkName, uplinkUUID, eeName, eeNamespace, ips)
```
It will be used by another PR for windows. That's why I make it public here.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,751 @@
+		attachInfo[ovsExternalIDEntityNamespace] = entityNamespace
+	}
+	if len(ips) != 0 {
+		attachInfo[ovsExternalIDIPs] = strings.Join(ips, ipsSplitter)
```
The ovsExternalIDIPs is set according to ExternalNode.interfaces[0].IPs. We don't require it is consistent with the IPs configured on the link, e.g. there are multiple IPs on the link, but user only configures one of them for ExternalNode.interfaces[0]. If we do not cache the IPs, it will cause the InterfaceConfig for the ExternalEntity which is loaded from OVSDB has no IPs, and we will need to wait ExternalNodeController reconciliation process to set IPs. Do you prefer to remove ovsExternalIDIPs from attchInfo?

mengdie-song replied:
attachInfo here is the same as externalIDs. They are consistent.

tnqn replied:
I meant https://github.com/antrea-io/antrea/blob/684dca3e30a333ff0653ea54df4060ae9e13406e/pkg/agent/cniserver/pod_configuration.go#L50

mengdie-song replied:
Got it, done.

mengdie-song commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
+	return false
+}
+
+func GetUplinkConfig(uplinkIfName string) (*config.AdapterNetConfig, error) {
```
I have refactored the this util file according to the comments. Please help review the latest version, thanks a lot.

mengdie-song commented on pkg/agent/util/net_linux.go:
```diff
@@ -237,3 +242,170 @@ func DeleteOVSPort(brName, portName string) error {
+		MAC:     iface.HardwareAddr,
+		IPs:     addrs,
+		Routes:  routes,
+		Gateway: gw,
```
Right, it is not used by ExternalNodeController right now. Removed it.

jianjuns commented on build/yamls/externalnode/vm-agent-rbac.yml:
```diff
@@ -0,0 +1,92 @@
+  - apiGroups:
+      - crd.antrea.io
+    resources:
+      - externalnodes
```
Can we restrict accesses to some resources to `vm-ns`?

mengdie-song replied:
I think the only resource we need to restrict is externalnodes. Since we can not specify namespace for ClusterRole, I have changed this part by adding a new Role for externalnodes and binding it to vm-agent.

tnqn commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,765 @@
+	if err := c.ofClient.UninstallVMUplinkFlows(portName); err != nil {
+		return fmt.Errorf("failed to uninstall uplink and host port openflow entries, portName %s, err %v", portName, err)
+	}
+	klog.InfoS("Removed the flows installed to forward packet between uplinkPort and hostPort.", "hostInterface", portName)
```
```suggestion
	klog.InfoS("Removed the flows installed to forward packet between uplinkPort and hostPort", "hostInterface", portName)
```

tnqn commented on pkg/agent/openflow/client.go:
```diff
@@ -337,7 +337,7 @@ type Client interface {
 	// or ip, port, protocol and direction. It is used to bypass NetworkPolicy enforcement on a VM for the particular
 	// traffic.
-	InstallPolicyBypassFlows(protocol binding.Protocol, ipnet *net.IPNet, ip net.IP, port uint16, isIngress bool) error
+	InstallPolicyBypassFlows(protocol binding.Protocol, ipnet *net.IPNet, port uint16, isIngress bool) error
```
```suggestion
	InstallPolicyBypassFlows(protocol binding.Protocol, ipNet *net.IPNet, port uint16, isIngress bool) error
```

tnqn commented on pkg/agent/openflow/externalnode_connectivity.go:
```diff
@@ -152,7 +152,7 @@ func (f *featureExternalNodeConnectivity) replayFlows() []binding.Flow {
 }
 
-func (f *featureExternalNodeConnectivity) policyBypassFlow(protocol binding.Protocol, ipnet *net.IPNet, ip net.IP, port uint16, isIngress bool) binding.Flow {
+func (f *featureExternalNodeConnectivity) policyBypassFlow(protocol binding.Protocol, ipnet *net.IPNet, port uint16, isIngress bool) binding.Flow {
```
ditto

tnqn commented on pkg/agent/util/net.go:
```diff
@@ -415,3 +415,17 @@ func GenerateRandomMAC() net.HardwareAddr {
 	return buf
 }
+
+func GetGlobalIPNetsByName(link *net.Interface) ([]*net.IPNet, error) {
```
The function doesn't get IPs "by name". And I don't quite get what "global" means here. If it means "scope global", it doesn't do any check.

mengdie-song replied:
Has changed to GetIPNetsByLink.

tnqn commented on pkg/ovs/ovsctl/appctl.go:
```diff
@@ -227,3 +228,49 @@ func (c *ovsCtlClient) GetDPFeatures() (map[DPFeature]bool, error) {
+		line := scanner.Text()
+		fields := strings.Split(line, ": ")
+		if len(fields) != 2 {
+			klog.InfoS("Unexpected output from dpif/show ovs-system", "line", line)
```
this may mismatch other cases, for example `port 1: genev_sys_6081 (geneve: packet_type=ptap)`. Although ExternalNode doesn't have such port, this is a generic function so should be more precise.
```
system@ovs-system:
  lookups: hit:188080 missed:6956 lost:0
  flows: 0
  masks: hit:1111386 total:0 hit/pkt:5.70
  port 0: ovs-system (internal)
  port 1: genev_sys_6081 (geneve: packet_type=ptap)
  port 2: antrea-gw0 (internal)
  port 3: nginx-76-8686ea
  port 4: nginx-76-92d4d0
```

mengdie-song replied:
Thanks for pointing out. I have changed this part

tnqn commented on pkg/ovs/ovsctl/appctl.go:
```diff
@@ -227,3 +228,49 @@ func (c *ovsCtlClient) GetDPFeatures() (map[DPFeature]bool, error) {
+		}
+		nameStr := fields[1]
+		ifName := strings.Split(nameStr, "(internal)")[0]
+		if ifName != name {
```
there is a space before `(internal)`, so the name would never match if it tries to remove an internal port.
```
port 2: antrea-gw0 (internal)
```

mengdie-song replied:
Done

tnqn commented on build/yamls/externalnode/vm-agent-rbac.yml:
```diff
@@ -0,0 +1,113 @@
+  - kind: ServiceAccount
+    name: vm-agent
+    namespace: vm-ns # Change the Namespace to where vm-agent is expected to run.
+
```
there are two empty lines here, remove one

tnqn commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,765 @@
+}
+
+func (c *ExternalNodeController) enqueueExternalNodeUpdate(oldObj interface{}, newObj interface{}) {
+	oldEn := oldObj.(*v1alpha1.ExternalNode)
```
```suggestion
	oldEN := oldObj.(*v1alpha1.ExternalNode)
```

mengdie-song replied:
done

tnqn commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,765 @@
+
+func (c *ExternalNodeController) reconcilePolicyBypassFlows() error {
+	rules := c.getPolicyBypassRules()
+	for _, rhp := range rules {
```
what's rhp short for?

mengdie-song replied:
Have changed to rule here.

mengdie-song commented on pkg/agent/controller/networkpolicy/reject.go:
```diff
@@ -283,6 +294,9 @@ func getRejectPacketOutMutateFunc(rejectType RejectType) func(binding.PacketOutB
 	case RejectLocalToRemote:
 		tableID := openflow.L3ForwardingTable.GetID()
+		if nodeType == config.ExternalNode {
+			tableID = openflow.IngressSecurityClassifierTable.GetID()
```
Sure, I have updated this part.

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,765 @@
+)
+
+var (
+	keyFunc      = cache.DeletionHandlingMetaNamespaceKeyFunc
```
MetaNamespaceKeyFunc should be good enough, updated.

wenyingd commented on pkg/agent/externalnode/external_node_controller_linux.go:
```diff
@@ -0,0 +1,61 @@
+	if err != nil {
+		return err
+	}
+	fromExists := true
```
But we don't need to check the existstence of source interface any more. We can decide it directly with the parameter. For deletion case, we just do not consume the source interface name.

mengdie-song replied:
I have updated accordingly, please help check the latest version, thanks.

wenyingd reviewed (commented):
Overall LGTM.

tnqn commented on pkg/ovs/ovsctl/appctl.go:
```diff
@@ -227,3 +228,35 @@ func (c *ovsCtlClient) GetDPFeatures() (map[DPFeature]bool, error) {
+	for scanner.Scan() {
+		line := scanner.Text()
+		fields := strings.Split(line, ": ")
+		nameStr := fields[1]
```
does it work with the first line `system@ovs-system:`? It panics according to my test. https://go.dev/play/p/Ow94O1eg6g0

mengdie-song replied:
Thanks for pointing out, done

mengdie-song commented on cmd/antrea-agent/agent.go:
```diff
@@ -611,6 +654,10 @@ func run(o *Options) error {
 	}
 
+	if localExternalNodeInformer != nil {
+		go localExternalNodeInformer.Run(stopCh)
```
done

mengdie-song commented on pkg/agent/controller/networkpolicy/cache.go:
```diff
@@ -443,6 +452,24 @@ func (c *ruleCache) processPodUpdate(e interface{}) {
+				c.onAppliedToGroupUpdate(group)
+				break
+			}
+		}
```
I have changed controller part of code as suggested. Could you help review that part, thanks!

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,774 @@
+	}
+	if err := wait.PollImmediateUntil(5*time.Second, func() (done bool, err error) {
+		if err = c.reconcile(); err != nil {
+			return false, nil
```
Thanks for the details. I have updated this part.

tnqn reviewed (commented):
LGTM, one minor comment, and the commit body is not aligned properly

mengdie-song commented on pkg/agent/externalnode/external_node_controller.go:
```diff
@@ -0,0 +1,682 @@
+}
+
+func (c *ExternalNodeController) reconcile() error {
+	klog.InfoS("Reconciliation for controller", "name", controllerName)
```
Thanks, done.

wenyingd reviewed (commented):
LGTM

tnqn commented:
@mengdie-song where is this change?
> 5. Update NetworkPolicy reconciler to invoke GetInterfacesByEntity()
> and GetContainerInterfacesByPod() for ExternalEntity and
> Pod separately.

mengdie-song commented:
@tnqn  Thanks a lot for the careful review. I realized that we have moved this part into a separate PR https://github.com/antrea-io/antrea/pull/3768 which has already been merged. I have updated the commit message.

tnqn reviewed (approved):
LGTM

tnqn commented:
Go / Verify docs and spelling (pull_request) Failing after 17s — Verify docs and spelling
```
Checking table of contents are up to date...
2022/08/02 04:03:12 docs/vm-installation.md: missing opening TOC tag
```

wenyingd commented:
> Go / Verify docs and spelling (pull_request) Failing after 17s — Verify docs and spelling
> 
> ```
> Checking table of contents are up to date...
> 2022/08/02 04:03:12 docs/vm-installation.md: missing opening TOC tag
> ```

This will be resolved in another PR for documentation: https://github.com/antrea-io/antrea/pull/3963

wenyingd commented:
/test-all
